File: day7.lua
#!/usr/bin/env lua

local function nextpow10(x)
    local p = 10
    while p <= x do p = p * 10 end
    return p
end

local function solve(terms, test, base)
    local sum = 0
    local p = base^#terms - 1
    for i = 0, p do
        local accum = terms[1]
        local quot = i
        for j = 2, #terms do
            local op = quot % base
            quot = quot // base
            if op == 0 then accum = accum * terms[j]
            elseif op == 1 then accum = accum + terms[j]
            elseif op == 2 then accum = accum * nextpow10(terms[j]) + terms[j] end
            if accum > test then break end
        end
        if accum == test then
            sum = sum + test
            break
        end
    end
    return sum
end

local file = io.open("day7.txt")
local sum = { 0, 0 }

for line in file:lines() do
    local test = tonumber(line:match("^(%d+):"))
    local terms = {}
    for term in line:gmatch(" (%d+)") do
        table.insert(terms, tonumber(term))
    end
    sum[1] = sum[1] + solve(terms, test, 2)
    sum[2] = sum[2] + solve(terms, test, 3)
end

print(sum[1], sum[2])

File: day7.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

int nextpow10(int x)
{
    int p = 10;
    while (p <= x) p *= 10;
    return p;
}

long long solve(const int* terms, int n, long long test, int base)
{
    long long sum = 0;
    int p = pow(base, n);
    for (int i = 0; i < p; i++) {
        long long accum = terms[0];
        int quot = i;
        for (int j = 1; j < n; j++) {
            int op = quot % base;
            quot /= base;
            switch (op) {
            case 0: accum *= terms[j]; break;
            case 1: accum += terms[j]; break;
            case 2: accum = accum * nextpow10(terms[j]) + terms[j]; break;
            }
            if (accum > test) break;
        }
        if (accum == test) {
            sum += test;
            break;
        }
    }
    return sum;
}

int main()
{
    FILE* file = fopen("day7.txt", "r");
    long long sum = 0, sum2 = 0;
    char line[256];

    while (fgets(line, sizeof(line), file)) {
        long long test;
        char* t = strtok(line, ":");
        sscanf(t, "%lld", &test);

        int terms[16], n = 0;
        while ((t = strtok(NULL, " \n"))) {
            sscanf(t, "%d", &terms[n++]);
        }

        sum += solve(terms, n, test, 2);
        sum2 += solve(terms, n, test, 3);
    }

    printf("%lld %lld\n", sum, sum2);
}
